### PR TITLE
Gv 33/error handling

### PR DESCRIPTION
**Pull request by Mehul-Gandhi**
### Jira Ticket

[Jira Ticket](https://dashbd.atlassian.net/jira/software/projects/GV/boards/4?selectedIssue=GV-33)

### Description

Create new HTTP Status Codes for backend APIs for Errors

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would change existing functionality)

### Changes

Add 4xx and 5xx responses for API calls that return an error.

### Testing

`http://localhost/api/v2/students/johnsmith@berkeley.edu/projections`
Status 401
image

`http://localhost/api/v2/students/gandhi@berkeley.edu/projections`
(In database) Status 200
image
```
{
  "zeros": 220,
  "pace": 500,
  "perfect": 500
}
```
`http://localhost/api/v2/students/notInDatabase@berkeley.edu/projections`
404
image

`http://localhost/api/v2/isadmin`
(gandhi@berkeley.edu)
image

After taking myself off of config.json
`http://localhost/api/v2/isadmin`
(gandhi@berkeley.edu)
image

http://localhost/api/v2/students/fake@gmail.com/grades
Retrieve grades as a not admin person for another student.
image

http://localhost/api/v2/students/fake@gmail.com/grades
As an admin, 404 not found error.
image

http://localhost/api/v2/students/abcd@berkeley.edu/grades
Student in spreadsheet, Status 200.
image

`http://localhost/api/v2/students/abcd@berkeley.edu/progressquerystring`
Student in spreadsheet, status 200
image

`http://localhost/api/v2/students/abcde@berkeley.edu/progressquerystring`
Student not in spreadsheet. Status 404.
image

Bins
`http://localhost/api/v2/bins`
image

Suggestion: Make a new ticket for `progressquerystring` because for status 200, a JSON is not returned, but for errors, a JSON is returned.

### Checklist

- I cannot change the branch name to GV-33/* because of branch protection rules. I cannot push to GV-33 because of branch protection rules. Therefore, I named this ticket GV-333.
- [ ] My branch name matches the format: `<ticket-id>/<brief-description-of-change>`
- [ ] My PR name matches the format: `[<ticket-id>] <brief-description-of-change>`
- [x] I have added doc-comments to all new functions ([JSDoc](https://jsdoc.app/) for JS and [Docstrings](https://peps.python.org/pep-0257/) for Python)
- [x] I have reviewed all of my code
- [x] My code only contains major changes related to my ticket

### CODE REVIEW

**Comment by Mehul-Gandhi** [comment]
I think `switch (typeof error)` compares by error objects and not strings, so I switched it back to `switch (err.name)`. All error cases were going to 5xx and not 5xx in my switch statements.

File: api/v2/Routes/students/index.js [hunk 1]
```diff
@@ -25,8 +25,20 @@ router.use('/:email/projections', ProjectionsRouter);
...
         return res.status(200).json({ students });
+    } catch (err) {
+        switch (err.name) {
+            case 'StudentNotEnrolledError':
+            case 'KeyNotFoundError':
+                console.error(`Error fetching student with id ${id}`, err);
+                return res.status(404).json({ message: "Error fetching student."});
+            default:
+                console.error(`Internal service error fetching student with id ${id}`, err);
+                return res.status(500).json({ message: "Internal server error." });
```
Comment (Connor-Bernard) [inline api/v2/Routes/students/index.js]: same error as thrown by codeql

**Review by Connor-Bernard — CHANGES_REQUESTED** [review]

File: api/v2/Routes/isadmin/index.js [hunk 2]
```diff
@@ -6,13 +6,24 @@ const router = Router({ mergeParams: true });
...
+        const adminStatus = await isAdmin(authEmail);
+        return res.status(200).json({ isAdmin: adminStatus });
+    } catch (err) {
+        switch (err.name) {
+            case 'AuthorizationError':
+                console.error('AuthorizationError:', err);
+                return res.status(401).json({ message: err.message });
+            default:
+                console.error('Internal Server Error:', err);
+                return res.status(500).json({ message: 'Internal Server Error' });
```
Comment (Connor-Bernard) [inline api/v2/Routes/isadmin/index.js:24]: Same comment as before, but for the sake of time and getting this ticket completed, please just open a ticket to refactor the individual try catches to a generic error handling middleware

File: api/v2/Routes/students/grades/index.js [hunk 3]
```diff
@@ -8,18 +8,30 @@ import { isAdmin } from '../../../../lib/userlib.mjs';
 const router = Router({ mergeParams: true });
 
 router.get('/', async (req, res) => {
-    const { id } = req.params;
-    const maxScores = await getMaxScores();
+    const { id } = req.params; // the id is the student's email
     let studentScores;
```
Comment (Connor-Bernard) [inline api/v2/Routes/students/grades/index.js]: Move this inside the try block if not otherwise referenced in this scope.

**Review by Connor-Bernard — CHANGES_REQUESTED** [review]
Looks a lot better.  Just one quick fix

File: api/v2/Routes/students/grades/index.js (hunk 3 above)
Comment (Connor-Bernard) [reply api/v2/Routes/students/grades/index.js]: resolved in [f6beb91](https://github.com/AFA-Tooling/GradeView/pull/12/commits/f6beb91531116dbd924697e08961e66d79b0e17a)

**Review by Connor-Bernard — APPROVED** [review]
LGTM

**Pull request merged**